docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,489 @@
+
+The disadvantage of using containers built `FROM scratch` is the lack of system
+binaries provided a base Operation System image makes it difficult to
+troubleshoot running containers. Kubernetes should enable troubleshooting pods
+regardless of the contents of the container image.
+
+## Goals and Non-Goals
+
+Goals include:
+
+*   Enable troubleshooting of minimal container images
+*   Allow troubleshooting of containers in `CrashLoopBackoff`
+*   Improve supportability of native Kubernetes applications
+*   Enable novel uses of short-lived containers in pods
+
+Non-Goals of this proposal are:
+
+*   Fully design the ability to cross-mount container image filesystems. This is
+    complex and should be covered in a separate doc.
+*   Guarantee resources for ad-hoc troubleshooting. If troubleshooting causes a
+    pod to exceed its resource limit it may be evicted.
+
+## Requirements
+
+A solution to troubleshoot minimal container images MUST:
+
+*   troubleshoot arbitrary running containers with minimal prior configuration
+*   allow access to all namespaces shared by containers in a pod and the mount
+    namespaces of individual containers
+*   fetch troubleshooting utilities at debug time rather than at the time of pod
+    initialization
+*   respect admission restrictions
+*   allow introspection of pod state using existing tools (no hidden containers)
+*   support arbitrary runtimes via the CRI (possibly with reduced feature set)
+
+A good solution SHOULD:
+
+*   have an excellent user experience (i.e. should be a feature of the platform
+    rather than a config-time solution)
+*   require no direct access to the node
+*   have no inherent side effects to the running container image
+*   re-use existing container image distribution channels
+*   enable detection of pods that have been modified by an exec ("tainted" pods)
+
+## kubectl debug
+
+We will introduce an "execute a debug container in a pod" pattern that parallels
+`kubectl exec`'s "execute a debug process in a container" in a new command,
+`kubectl debug`, which has two modes of operation:
+
+1.  *Copy Debug* mode creates a copy of a pod with minor changes to the pod spec
+    specified on the command line.
+1.  *Running Debug* mode causes Kubernetes to run a new *debug container* in the
+    pod context. A *debug container* is not part of the pod spec and cannot be
+    configured by the user. It is created by a debug operation on an existing
+    pod and reported in `PodStatus`.
+
+The mode distinction is for the purposes of this proposal, and the end user need
+not be aware of them. They form a single debugging command that allows the user
+to modify a running pod if possible and create a copy otherwise. The
+determination is made based on container name, for example:
+
+```
+% kubectl run example --image=nginx --restart=Never
+pod "example" created
+
+# Container "shell" does not exist, add a new container to running pod
+kubectl debug -p example -c shell --image=debian
```

Comment:
I agree with your analysis and like your suggestions better than the original. I especially like having a consistent debug "target" specified as an argument rather than a flag.

One correction for your example, it's not `--in-place` and `--command` that are mutually exclusive, it's that `--in-place` cannot be used with a container name that's already in the pod spec. Also, I agree that using `-p/--pod` to refer to a pod that doesn't exist is inconsistent with the rest of `kubectl`. Maybe `--copy-name`?

One nit is that `--in-place` might imply one can edit the pod spec in place, which isn't true, but I can't think of a better way to convey the semantics. Just brainstorming some things:

```
# ERROR: container  'nginx' already exists
kubectl debug nginx --in-place -c nginx --command -- /bin/sh

# Running debug adding a new container with generated defaults for everything, including image.
# Is a default image a bad idea?
kubectl debug nginx --in-place

# If specifying an image is required then we could make that the argument that activates running debug
kubectl debug nginx --run busybox
# ERROR: awkward side effect of above
kubectl debug nginx --run busybox --image busybox

# What would we want this to do?
kubectl debug nginx
```